source/pervasive/nonlinear_arith/div.rs

```diff
@@ -0,0 +1,1088 @@
+    {
+    lemma_div_pos_is_pos(x, y);
+    lemma_div_is_ordered_by_denominator(x, 1, y);
+    };
+}
+
+/// if a dividend is a whole number and the divisor is a natural number and their
+/// quotient is 0, this implies that the dividend is smaller than the divisor
+// #[verifier::spinoff_prover]
+pub proof fn lemma_small_div_converse_auto()
+    ensures forall |x: int, d:int| 0 <= x && 0 < d && #[trigger](x / d) == 0 ==> x < d,
+{
+    assert forall |x: int, d: int| 0 <= x && 0 < d &&  #[trigger](x / d) == 0 implies x < d by
+    {
+        lemma_div_induction_auto(d, x, |u: int| 0 <= u && 0 < d && u / d == 0 ==> u < d); 
+    }
+}
+
+/// Division of a positive integer by a positive integer less than or equal to it is nonzero
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_non_zero(x: int, d: int)
+    requires x >= d > 0
+    ensures x / d > 0
+{
+    lemma_div_pos_is_pos_auto();
+    if x / d == 0 {
+    lemma_small_div_converse_auto();
+    }
+}
+
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_non_zero_auto()
+    ensures forall |x: int, d: int| x >= d > 0 ==> #[trigger] (x / d) > 0
+{
+    assert forall |x: int, d: int| x >= d > 0 implies #[trigger] (x / d) > 0 by
+    {
+        lemma_div_non_zero(x, d);
+    }
+}
+
+spec fn div (x:int, y: int) -> int
+    recommends y != 0
+{
+    x / y
+}
+
+/// given two fractions with the same numerator, the order of numbers is determined by 
+/// the denominators. However, if the numerator is 0, the fractions are equal regardless of 
+/// the denominators' values
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_is_ordered_by_denominator(x: int, y: int, z: int)
+    requires 
+        0 <= x,
+        1 <= y <= z
+    ensures 
+        x / y >= x / z
+    decreases x
+{
+    reveal(div_recursive);
+    reveal(div_pos);
+    lemma_div_is_div_recursive_auto();
+
+    assert(forall |u: int, d: int| d > 0 ==> #[trigger]div_recursive(u, d) == #[trigger](div(u, d)));
+
+    if (x < z)
+    {
+        lemma_div_is_ordered(0, x, y);
+    }
+    else
+    {
+        lemma_div_is_ordered(x - z, x - y, y);
+        lemma_div_is_ordered_by_denominator(x - z, y, z);
+    }
+  
+}
+
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_is_ordered_by_denominator_auto()
+    ensures forall |x: int, y: int, z: int| 0 <= x && 1 <= y <= z ==> #[trigger](x / y) >= #[trigger](x / z)
+{
+    assert forall |x: int, y: int, z: int| 0 <= x && 1 <= y <= z implies #[trigger](x / y) >= #[trigger](x / z) by
+    {
+        lemma_div_is_ordered_by_denominator(x, y, z);
+    }
+}
+
+/// given two fractions with the same numerator, the order of numbers is strictly determined by 
+/// the denominators.
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_is_strictly_ordered_by_denominator(x: int, d: int)
+    requires 
+        0 < x, 
+        1 < d
+    ensures 
+        x / d  < x
+    decreases x
+{
+    lemma_div_induction_auto(d, x, |u: int| 0 < u ==> u / d < u);
+}
```

Comment:
This comment is not what this lemmas is showing though, is it? It is showing that a number divided is strictly smaller than the number itself.